src/ILVerify/src/Verifier.cs

```diff
@@ -0,0 +1,224 @@
+using System.Reflection.PortableExecutable;
+using System.Resources;
+using System.Text;
+using Internal.IL;
+using Internal.TypeSystem;
+using Internal.TypeSystem.Ecma;
+
+namespace ILVerify
+{
+    public delegate bool ShouldVerifyMethod(string name);
+
+    public interface IResolver
+    {
+        PEReader Resolve(AssemblyName name);
+    }
+
+    public class Verifier
+    {
+        private Lazy<ResourceManager> _stringResourceManager =
+            new Lazy<ResourceManager>(() => new ResourceManager("ILVerify.Resources.Strings", Assembly.GetExecutingAssembly()));
+
+        private SimpleTypeSystemContext _typeSystemContext;
+
+        public ShouldVerifyMethod ShouldVerifyMethod { private get; set; }
+
+        public Verifier(IResolver resolver)
+        {
+            _typeSystemContext = new SimpleTypeSystemContext(resolver);
+        }
+
+        internal Verifier(SimpleTypeSystemContext context)
+        {
+            _typeSystemContext = context;
+        }
+
+        public void SetSystemModuleName(AssemblyName name)
+        {
+            _typeSystemContext.SetSystemModule(_typeSystemContext.GetModule(name));
+        }
+
+        public VerificationResult Verify(AssemblyName moduleToVerify)
+        {
+            if (moduleToVerify == null)
+            {
+                throw new ArgumentNullException(nameof(moduleToVerify));
+            }
+
+            try
+            {
+                if (_typeSystemContext.SystemModule is null)
+                {
+                    return new VerificationResult() { NumErrors = 1, Message = "No system module specified" };
+                }
+
+                EcmaModule module = _typeSystemContext.GetModule(moduleToVerify);
+                return VerifyModule(module, moduleToVerify.Name);
+            }
+            catch (VerifierException e)
+            {
+                return new VerificationResult() { NumErrors = 1, Message = e.Message };
+            }
+        }
+
+        private VerificationResult VerifyModule(EcmaModule module, string path)
+        {
+            foreach (var methodHandle in module.MetadataReader.MethodDefinitions)
+            {
+                var method = (EcmaMethod)module.GetMethod(methodHandle);
+
+                var methodIL = EcmaMethodIL.Create(method);
+                if (methodIL == null)
+                {
+                    continue;
+                }
+
+                var methodName = method.ToString();
+                if (ShouldVerifyMethod != null && !ShouldVerifyMethod(methodName))
+                {
+                    continue;
+                }
+
+                var result = VerifyMethod(method, methodIL, path);
+                if (result.NumErrors > 0)
+                {
+                    return result;
+                }
+            }
+
+            return new VerificationResult();
+        }
+
+        internal VerificationResult VerifyMethod(MethodDesc method, MethodIL methodIL, string moduleName)
```

Comment:
Do these methods need to take `moduleName`? You should be always able to get the module name from `MethodDesc`.

Reply:
I should rename the `moduleName`. This parameter is used to maintain the old command-line output, which inclues file paths.
But when the verifier is used in-memory, I'm passing in the module name.